libs/media-player/src/lib/timeline-ruler.tsx

```diff
@@ -0,0 +1,154 @@
+import { isNull } from '@coscrad/validation-constraints';
+import { Box } from '@mui/material';
+import { useEffect } from 'react';
+import { RULER_TICKS_AND_NUMBERS_COLOR } from './ruler-tick';
+import { asFormattedMediaTimecodeString } from './shared/as-formatted-media-timecode-string';
+import {
+    MAX_CANVAS_WIDTH_IN_PIXELS,
+    RULER_TICK_WIDTH_IN_PIXELS,
+    ZoomLevels,
+} from './shared/constants';
+import { TimelineCanvas } from './timeline-canvas';
+
+const getNumberOfCanvases = (lastCanvasWidth: number, numberOfWholeCanvases: number) => {
+    if (lastCanvasWidth > 0) {
+        return numberOfWholeCanvases + 1;
+    } else {
+        return numberOfWholeCanvases;
+    }
+};
+
+export type TimelineTimecodeProps = {
+    currentSecond: number;
+    label?: string;
+};
+
+export type CanvasProps = {
+    canvasWidth: number;
+    ticksForThisCanvas: TimelineTimecodeProps[];
+};
+
+const sxProps = {
+    borderBottom: `1px solid ${RULER_TICKS_AND_NUMBERS_COLOR}`,
+};
+
+interface TimelineProps {
+    duration: number;
+    zoomLevelConfig: ZoomLevels;
+    timelineTrackHeight: number;
+    setRulerWidth: (rulerWidth: number) => void;
+}
+
+export const TimelineRuler = ({
+    duration,
+    zoomLevelConfig,
+    timelineTrackHeight,
+    setRulerWidth,
+}: TimelineProps): JSX.Element => {
+    const roundedSecondsInDuration: number = Math.ceil(duration);
+
+    const {
+        rulerTickXCoordinateOffset,
+        rulerTickFrequencyInSeconds,
+        timecodeDisplayFrequencyInSeconds,
+    } = zoomLevelConfig;
+
+    /**
+     * Create an array of all the ticks on the ruler from the seconds in the media duration
+     * based on the zoom level config e.g.,
+     * {
+     *      zoomLevel: 0,
+     *      rulerTickXCoordinateOffset: 40,  // 40px between ticks at this zoom level
+     *      rulerTickFrequencyInSeconds: 1,  // a tick for each second, or less as the zoom decreases
+     *      timecodeDisplayFrequencyInSeconds: 10  // display the timecode every 10th second in the timeline
+     * }
+     */
+    const ticksOnTimelineAtCurrentZoom = [...Array(roundedSecondsInDuration).keys()]
+        .map((second) => {
+            if (second % rulerTickFrequencyInSeconds === 0) {
+                if (second % timecodeDisplayFrequencyInSeconds === 0) {
+                    const timeCode = asFormattedMediaTimecodeString(second);
+
+                    return {
+                        second: second,
+                        label: timeCode,
+                    };
+                }
+
+                return {
+                    second: second,
+                };
+            }
+
+            return null;
+        })
+        .filter((second) => !isNull(second)) as unknown as TimelineTimecodeProps[];
```

Comment:
I think you can avoid the `.filter` if you use `.flatMap` and return `[]` where you returned null previously, and otherwise `[{...}`.